cloudscheduler/openstackcluster.py

```diff
@@ -238,14 +241,52 @@ def vm_create(self, vm_name, vm_type, vm_user, vm_networkassoc,
                     netid = []
         else:
             netid = []
-        # Need to get the rotating hostname from the google code to use for here.  
+        # Need to get the rotating hostname from the google code to use for here.
         name = self._generate_next_name()
         instance = None
 
         if name:
+            log.info("Trying to create VM on %s: " % self.name)
             try:
-                instance = nova.servers.create(name=name, image=imageobj, flavor=flavor, key_name=key_name, 
-                                               availability_zone=self.placement_zone, nics =netid, userdata=user_data, security_groups=sec_group)
+                if not boot_volume:
+                    instance = nova.servers.create(name=name,
+                                                   image=imageobj,
+                                                   flavor=flavor,
+                                                   key_name=key_name,
+                                                   availability_zone=self.placement_zone,
+                                                   nics =netid,
+                                                   userdata=user_data,
+                                                   security_groups=sec_group)
+                else:
+                    bdm = None
+                    log.debug("creating boot volume")
+                    try:
+                        bv_name = "vol-{}".format(name)
+                        if boot_volume_gb_per_core:
+                            bv_size = boot_volume_gb_per_core * cpu_cores
+                        else:
+                            bv_size = 20
+                        cv = cinder.volumes.create(name=bv_name,
+                                                   size=bv_size,
+                                                   imageRef=imageobj.id)
+                        while (cv.status != 'available'):
+                            time.sleep(1)
+                        cv = cinder.volumes.get(cv.id)
+                        cinder.volumes.set_bootable(cv, True)
+                        bdm = {'vda': str(cv.id) + ':::1'}
+                    except Exception as e:
+                        log.error("failed to create boot volume: {}".format(e))
+                        raise e
```

Comment:
For V2 we should also think about possible failures, e.g. during testing volumes were created fine, but creation of VMs failed e.g. due to wrong networks. This left the volume lying around and needed to be removed by hand. There's several ways to do it, IMHO but at this point in time nothing needed for V1.

Reply:
So I have added something to the exception handling in `vm_create`, but to add some cleanup to the vm deletion means we have to change the database model for the VMs. Note sure how significant an edit that is in the end.